text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,1062 @@
+   type Item = &'a T;
+   ...
+}
+```
+
+The problem here is that for `&'a T` to be well-formed, `T: 'a` must
+hold, but that is not specified in the where clauses. This RFC
+proposes using implied bounds to address this concern -- specifically,
+every impl is permitted to assume that all types which appear in the
+impl header (trait reference) are well-formed, and in turn each "user"
+of an impl will validate this requirement whenever they project out of
+a trait reference (e.g., to do a method call, or normalize an
+associated type).
+
+# Detailed design
+
+This section dives into detail on the proposed type rules.
+
+### A little type grammar
+
+We extend the type grammar from [RFC 192] with projections and slice
+types:
+
+    T = scalar (i32, u32, ...)        // Boring stuff
+      | X                             // Type variable
+      | Id<P0..Pn>                    // Nominal type (struct, enum)
+      | &r T                          // Reference (mut doesn't matter here)
+      | O0..On+r                      // Object type
+      | [T]                           // Slice type
+      | for<r..> fn(T1..Tn) -> T0     // Function pointer
+      | <P0 as Trait<P1..Pn>>::Id     // Projection
+    P = r                             // Region name
+      | T                             // Type
+    O = for<r..> TraitId<P1..Pn>      // Object type fragment
+    r = 'x                            // Region name
+    
+We'll use this to describe the rules in detail.
+    
+### Syntactic definition of the outlives relation
+
+The outlives relation is defined in purely syntactic terms as follows.
+These are inference rules written in a primitive ASCII notation. :) As
+part of defining the outlives relation, we need to track the set of
+lifetimes that are bound within the type we are looking at.  Let's
+call that set `R=<r0..rn>`. Initially, this set `R` is empty, but it
+will grow as we traverse through types like fns or object fragments,
+which can bind region names via `for<..>`.
+
+#### Simple outlives rules
+
+Here are the rules covering the simple cases, where no type parameters
+or projections are involved:
+
+    OutlivesScalar:
+      --------------------------------------------------
+      R ⊢ scalar: 'a
+
+    OutlivesNominalType:
+      ∀i. R ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ Id<P0..Pn>: 'a
+
+    OutlivesReference:
+      R ⊢ 'x: 'a
+      R ⊢ T: 'a
+      --------------------------------------------------
+      R ⊢ &'x T: 'a
+
+    OutlivesObject:
+      ∀i. R ⊢ Oi: 'a
+      R ⊢ 'x: 'a
+      --------------------------------------------------
+      R ⊢ O0..On+'x: 'a
+
+    OutlivesFunction:
+      ∀i. R,r.. ⊢ Ti: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0
+
+    OutlivesTraitRef:
+      ∀i. R,r.. ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> TraitId<P0..Pn>: 'a      
```

Comment:
A trait reference does not yet give a self type, does it? If I understood this correctly, a trait reference is what I write in places like "where T: Trait<...>".
Grammatically, the two look exactly equal: What you have in this "OutlivesTraitRef" rule is grammatically also a valid object type fragment, according to the grammar above. If the two are different, object references should be added to the grammar somewhere - and there should be some way to tell the two apart when they are mentioned in this document.

Also, the subsection "WF checking a trait reference" gives only one rule, called "WfObjectFragment". How does that make sense?

And finally, I cannot find the outlives rule for object fragments anywhere, or the WF rule for trait references.